homeassistant/components/media_player/songpal.py

```diff
@@ -114,13 +125,80 @@ def __init__(self, name, endpoint):
 
-        self._sources = []
+        self._active_source = None
+        self._sources = {}
+
+    @property
+    def should_poll(self):
+        """Return True if the device should be polled."""
+        return self._poll
 
     async def initialize(self):
         """Initialize the device."""
         await self.dev.get_supported_methods()
         self._sysinfo = await self.dev.get_system_info()
 
+    async def async_activate_websocket(self):
+        """Activate websocket for listening if wanted."""
+        _LOGGER.info("Activating websocket connection..")
+        from songpal import (VolumeChange, ContentChange,
+                             PowerChange, ConnectChange)
+
+        async def _volume_changed(volume: VolumeChange):
+            _LOGGER.debug("Volume changed: %s", volume)
+            self._volume = volume.volume
+            self._is_muted = volume.mute
+            await self.async_update_ha_state()
+
+        async def _source_changed(content: ContentChange):
+            _LOGGER.debug("Source changed: %s", content)
+            if content.is_input:
+                self._active_source = self._sources[content.source]
+                _LOGGER.debug("New active source: %s", self._active_source)
+                await self.async_update_ha_state()
+            else:
+                _LOGGER.warning("Got non-handled content change: %s",
+                                content)
+
+        async def _power_changed(power: PowerChange):
+            _LOGGER.debug("Power changed: %s", power)
+            self._state = power.status
+            await self.async_update_ha_state()
+
+        async def _try_reconnect(x: ConnectChange):
+            _LOGGER.error("Got disconnected with %s, trying to reconnect.",
+                          x.exception)
+            self._available = False
+            self.dev.clear_notification_callbacks()
+            await self.async_update_ha_state()
+
+            # Try to reconnect forever, a successful reconnect will initialize
+            # the websocket connection again.
+            while not self._available:
+                await asyncio.sleep(10)
+                await self.async_update()
+                # We need to inform HA about the state in case we are coming
+                # back from a disconnected state.
+                await self.async_update_ha_state()
+
+        self.dev.on_notification(VolumeChange, _volume_changed)
+        self.dev.on_notification(ContentChange, _source_changed)
+        self.dev.on_notification(PowerChange, _power_changed)
+        self.dev.on_notification(ConnectChange, _try_reconnect)
+
+        async def listen_events():
+            await self.dev.listen_notifications()
+
+        async def handle_stop(event):
+            nonlocal remove_hass_stop_listener
+            remove_hass_stop_listener = None
+            await self.dev.stop_listen_notifications()
+
+        remove_hass_stop_listener = self.hass.bus.async_listen_once(
```

Comment:
We're not using the unsubscribe function, so we don't need to store that here. Just register the listener.